domain/application/state/unit.go

```diff
@@ -520,6 +525,170 @@ func (st *State) AddCAASUnits(
 	return errors.Capture(err)
 }
 
+// AddSubordinateUnit adds a unit to the specified subordinate application to
+// the application on the same machine as the given principal unit and records
+// the principal-subordinate relationship.
+//
+// The following error types can be expected:
+//   - [applicationerrors.ApplicationNotFound] when the subordinate application
+//     cannot be found.
+//   - [applicationerrors.UnitNotFound] when the unit cannot be found.
+func (st *State) AddSubordinateUnit(
+	ctx context.Context,
+	arg application.SubordinateUnitArg,
+) (coreunit.Name, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	var unitName coreunit.Name
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Check the application is a subordinate application.
+		if isSub, err := st.isSubordinate(ctx, tx, arg.SubordinateAppID); err != nil {
+			return errors.Errorf("checking is subordinate %q: %w", arg.SubordinateAppID, err)
+		} else if !isSub {
+			return applicationerrors.ApplicationNotSubordinate
+		}
+
+		// Check the application is alive.
+		if err := st.checkApplicationAlive(ctx, tx, arg.SubordinateAppID); err != nil {
+			return errors.Capture(err)
+		}
+
+		// Check this unit does not already have a subordinate unit from this
+		// application.
+		if err := st.checkNoSubordinateExists(ctx, tx, arg.SubordinateAppID, arg.PrincipalUnitName); err != nil {
+			return errors.Errorf("checking if subordinate already exists: %w", err)
+		}
+
+		// If on IAAS, place the subordinate on the same machine as the
+		// principal unit.
+		var placement deployment.Placement
+		if arg.ModelType == model.IAAS {
+			machineName, err := st.getUnitMachineName(ctx, tx, arg.PrincipalUnitName)
+			if err != nil {
+				return errors.Errorf("getting unit machine name: %w", err)
+			}
+			placement = deployment.Placement{
+				Type:      deployment.PlacementTypeMachine,
+				Directive: machineName.String(),
+			}
+		}
+
+		// Generate a new unit name.
+		unitName, err = st.newUnitName(ctx, tx, arg.SubordinateAppID)
+		if err != nil {
+			return errors.Errorf("getting new unit name for application %q: %w", arg.SubordinateAppID, err)
+		}
+
+		// Insert the new unit.
+		// TODO(storage) - read and use storage directives
+		insertArg := application.InsertUnitArg{
+			UnitName:         unitName,
+			Placement:        placement,
```

Comment:
Not your patch issue, but this is silly, we've got bifurcated methods, but they accept the same argument, one with a member that only applies to one method.